### PR TITLE
thunderbolt: set ports offline on host controller

### PR DESCRIPTION
**Pull request by krankun**
This change offline ports in the host controller
for retimers to enumerate in the NDA case.

-offline and rescan usb4 ports
-this enable enumeration of the retimers
-updates nvm to the enumerated retimers
-online usb4 port this will de-enumerate retimers

BUG=b:187506425
TEST=emerge-volteer fwupd

Change-Id: I1fde5675f27e91e18951b72f24fbec4113410807

Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [x] Feature
- [ ] Documentation

### CODE REVIEW

File: libfwupd/fwupd-enums.h [hunk 1]
```diff
@@ -493,6 +493,15 @@ typedef enum {
...
 #define FWUPD_DEVICE_FLAG_AFFECTS_FDE (1llu << 45)
+/**
+ *
+ * FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE :
+ *
+ * The device is not auto removed.
+ * Since 1.7.3
+ */
+#define FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE (1llu << 46)
+
```
Comment (superm1) [inline libfwupd/fwupd-enums.h:504]: @hughsie should these be internal flags perhaps?  I'm thinking yes.
@krankun IOW define them in `libfwupdplugin/fu-device.h`

File: libfwupd/fwupd-enums.c [hunk 2]
```diff
@@ -312,6 +314,8 @@ fwupd_device_flag_from_string(const gchar *device_flag)
 		return FWUPD_DEVICE_FLAG_UNREACHABLE;
 	if (g_strcmp0(device_flag, "affects-fde") == 0)
 		return FWUPD_DEVICE_FLAG_AFFECTS_FDE;
+	if (g_strcmp0(device_flag, "no-auto-remove") == 0)
+		return FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE;
```
Comment (superm1) [inline libfwupd/fwupd-enums.c:318]: if these end up being internal flags, then this should move to `libfwupdplugin/fu-device.c`

File: libfwupd/fwupd-enums.h (hunk 1 above)
Comment (hughsie) [reply libfwupd/fwupd-enums.h:504]: Yes, I think internal too -- they're of no use to the end user.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 3]
```diff
@@ -76,3 +76,32 @@ fu_plugin_init_vfuncs(FuPluginVfuncs *vfuncs)
...
+		}
+	}
+	return TRUE;
+}
+
+
+gboolean
+fu_plugin_composite_cleanup	(FuPlugin	*plugin,
+							 GPtrArray	*devices,
+							 GError		**error)
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c:98]: whitespace seems funny here, did clang do this?  It should be more like the composite_prepare function.

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 4]
```diff
@@ -76,3 +76,32 @@ fu_plugin_init_vfuncs(FuPluginVfuncs *vfuncs)
...
+{
+	for (guint i = 0; i < devices->len; i++) {
+		FuDevice *device = g_ptr_array_index (devices, i);
+		if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE)) {
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c:102]: I think you also need to check that the device comes from the thunderbolt plugin.  Composite prepare/composite cleanup could potentially have devices from other plugins.

File: plugins/thunderbolt/fu-self-test.c [hunk 5]
```diff
@@ -61,6 +61,28 @@ udev_mock_add_nvmem(UMockdevTestbed *bed, gboolean active, const char *parent, i
 	return path;
 }
 
+static gchar *
+udev_mock_add_usb4_port(UMockdevTestbed *bed, int id)
+{
+	gchar *path;
+	g_autofree gchar *name = NULL;
```
Comment (superm1) [inline plugins/thunderbolt/fu-self-test.c:68]: reverse xmas tree

File: plugins/thunderbolt/fu-self-test.c [hunk 6]
```diff
@@ -61,6 +61,28 @@ udev_mock_add_nvmem(UMockdevTestbed *bed, gboolean active, const char *parent, i
...
+
+	name = g_strdup_printf("usb4_port%d", id);
```
Comment (superm1) [inline plugins/thunderbolt/fu-self-test.c:70]: nit: just combine these to one line

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 7]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device)
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:324]: you should be passing in the GError **error as an argument here and then using that in this function rather than local errors.  The idea is supposed to be that if you set FALSE, then you need to return something in error.

The way this function is working right now, you're setting all those errors, but you're discarding them as they're local to the function.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 8]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+{
+	GError **error_local_offline = NULL;
+	GError **error_local_rescan = NULL;
+
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:328]: drop these, and just pass a single `**error` into the function.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 9]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+				    FWUPD_ERROR_INTERNAL,
+				    "Rescan on port failed");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device)
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:348]: pass an error into the function like the  `set_port_offline` one.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 10]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	if (udev_parent1 == NULL)
+		return NULL;
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	return fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+					       g_steal_pointer(&udev_parent2));
+}
+gboolean
+fu_thunderbolt_device_open(FuDevice *device, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = get_udev_parent_device(device);
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:383]: you need to be guarding for the NULL here and return FALSE in that case

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 11]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+
+static FuUdevDevice *
+get_udev_parent_device(FuDevice *device)
+{
+	g_autoptr(GUdevDevice) udev_parent1 = NULL;
+	g_autoptr(GUdevDevice) udev_parent2 = NULL;
+	GUdevDevice *udev_device = NULL;
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
+	if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER)
+		return NULL;
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:369]: error should be an argument to the futnction function and you need to set it if this was false

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 12]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	udev_parent1 = g_udev_device_get_parent(udev_device);
+	if (udev_parent1 == NULL)
+		return NULL;
+	udev_parent2 = g_udev_device_get_parent(udev_parent1);
+	return fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+					       g_steal_pointer(&udev_parent2));
+}
+gboolean
+fu_thunderbolt_device_open(FuDevice *device, GError **error)
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:381]: I'd suggest to rename it to `fu_thunderbolt_retimer_open` as this is only used in the retimer case

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 10 above)
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:383]: I think maybe an assertion makes sense here if this is not a retimer device then there was a programmer error

File: libfwupd/fwupd-enums.h (hunk 1 above)
Comment (superm1) [reply libfwupd/fwupd-enums.h:504]: Oh but they're used in `src/fu-udev-backend.c`, so maybe not.

File: src/fu-udev-backend.c [hunk 13]
```diff
@@ -41,9 +41,11 @@ fu_udev_backend_device_remove(FuUdevBackend *self, GUdevDevice *udev_device)
 	/* find the device we enumerated */
 	device_tmp =
 	    fu_backend_lookup_by_id(FU_BACKEND(self), g_udev_device_get_sysfs_path(udev_device));
-	if (device_tmp != NULL) {
-		if (g_getenv("FWUPD_PROBE_VERBOSE") != NULL) {
-			g_debug("UDEV %s removed", g_udev_device_get_sysfs_path(udev_device));
+	if (device_tmp != NULL &&
+	    fu_device_has_flag(device_tmp, FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE)) {
```
Comment (superm1) [inline src/fu-udev-backend.c:45]: shouldn't this be `!fu_device_has_flag(..)`?

**Review by superm1 — CHANGES_REQUESTED** [review]
pre-commit continues to fail and various nested comments.
No need to open a new PR each time.  You can just `git push --force`

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 14]
```diff
@@ -76,3 +76,32 @@ fu_plugin_init_vfuncs(FuPluginVfuncs *vfuncs)
 	vfuncs->device_registered = fu_plugin_thunderbolt_device_registered;
 	vfuncs->device_created = fu_plugin_thunderbolt_device_created;
 }
+
+gboolean
+fu_plugin_composite_prepare (FuPlugin *plugin,
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c:81]: On main the way this works, you need to declare the vfuncs for composite prepare and cleanup as `static` and then declare what they are in `fu_plugin_init_vfuncs`.  You can look at `plugins/dell-dock/fu-plugin-dell-dock.c` for an example of how this is supposed to be done.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 15]
```diff
@@ -320,6 +320,78 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device)
+{
+	GError **error_local_offline = NULL;
+	GError **error_local_rescan = NULL;
+
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/offline", "1", error_local_offline)) {
```
Comment (djcampello) [inline plugins/thunderbolt/fu-thunderbolt-device.c:329]: Are we hard coding usb4_port1?

Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c:329]: @djcampello  Yes currently there is just one port to update nvm and the other port shares the same so no need to update it , in later platforms EC needs to populate only the retimers which don't share the same NVM in that way we would grep the available usb4_portx and update nvm for all of them

File: src/fu-udev-backend.c (hunk 13 above)
Comment (krankun) [reply src/fu-udev-backend.c]: I have corrected it and got it functional with the new change but will keep this open until I debug and understand the implications if any

**Comment by krankun** [comment]
> pre-commit continues to fail and various nested comments. No need to open a new PR each time. You can just `git push --force`

lets see if the new force push fixes it

File: libfwupd/fwupd-enums.c (hunk 2 above)
Comment (krankun) [reply libfwupd/fwupd-enums.c:318]: let me look at these , still open

File: plugins/thunderbolt/fu-plugin-thunderbolt.c [hunk 16]
```diff
@@ -75,4 +101,6 @@ fu_plugin_init_vfuncs(FuPluginVfuncs *vfuncs)
 	vfuncs->startup = fu_plugin_thunderbolt_startup;
 	vfuncs->device_registered = fu_plugin_thunderbolt_device_registered;
 	vfuncs->device_created = fu_plugin_thunderbolt_device_created;
+	vfuncs->composite_prepare = fu_plugin_composite_prepare;
+	vfuncs->composite_cleanup = fu_plugin_composite_cleanup;
```
Comment (superm1) [inline plugins/thunderbolt/fu-plugin-thunderbolt.c:105]: you should include the word `thunderbolt` in these like other vfuncs to avoid potential symbol collisions when loaded into the daemon.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 17]
```diff
@@ -320,6 +320,105 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INTERNAL,
+				    "Rescan on port failed");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static void
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:344]: this should be type gboolean (and return TRUE/FALSE)

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 18]
```diff
@@ -320,6 +320,105 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	}
+	parent = get_udev_parent_device(device, error);
+	return fu_thunderbolt_device_set_port_offline(parent, error);
+}
+
+gboolean
+fu_thunderbolt_device_close(FuDevice *device, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = get_udev_parent_device(device, error);
+	fu_thunderbolt_device_set_port_online(parent, error);
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:418]: just return fu_thunderbolt_device_set_port_online(parent, error)

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 19]
```diff
@@ -320,6 +320,105 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+		return FALSE;
+	}
+	parent = get_udev_parent_device(device, error);
+	return fu_thunderbolt_device_set_port_offline(parent, error);
+}
+
+gboolean
+fu_thunderbolt_device_close(FuDevice *device, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = get_udev_parent_device(device, error);
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:417]: you need to test if `get_udev_parent_device` was NULL before passing it on

**Comment by krankun** [comment]
> pre-commit continues to fail and various nested comments. No need to open a new PR each time. You can just `git push --force`

I hope pre commit should pass now as there was a clang format issue with fu-thunderbolt-device.h

**Comment by krankun** [comment]
> > pre-commit continues to fail and various nested comments. No need to open a new PR each time. You can just `git push --force`
> 
> I hope pre commit should pass now as there was a clang format issue with fu-thunderbolt-device.h

precommit passed now with https://github.com/fwupd/fwupd/commit/b741625dcee3fdc6df5fe3d5c45e6d9cb24fc7c2

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 15 above)
Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: Closing this , Thank you!

File: src/fu-udev-backend.c (hunk 13 above)
Comment (krankun) [reply src/fu-udev-backend.c]: this is incorporated in latest

File: libfwupd/fwupd-enums.c (hunk 2 above)
Comment (krankun) [reply libfwupd/fwupd-enums.c:318]: should I move this flags to libfwupdplugin/fu-device.c ?

Comment (superm1) [reply libfwupd/fwupd-enums.c:318]: Yes, move it to fu-device.c and the matching one from fwupd-enums.h to fu-device.h

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 20]
```diff
@@ -320,6 +320,108 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
 	return (guint16)val;
 }
 
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device, GError **err)
+{
+	if (!fu_udev_device_write_sysfs(device, "usb4_port1/offline", "1", err)) {
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:326]: I think you might need an extra check that this file exists first.  Think about the case that this file isn't existing, this function should return true immediately right?

File: src/fu-util-common.c [hunk 21]
```diff
@@ -1217,6 +1217,10 @@ fu_util_device_flag_to_string(guint64 device_flag)
 		/* TRANSLATORS: we might ask the user the recovery key when next booting Windows */
 		return _("Full disk encryption secrets may be invalidated when updating");
 	}
+	if (device_flag == FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE) {
```
Comment (superm1) [inline src/fu-util-common.c:1220]: once you move the flag to an internal flag the changes for this file should go away

**Review by superm1 — COMMENTED** [review]
getting much closer, just move some stuff around and a minor nit left

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 20 above)
Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c:326]: as long as the usb4 port exists and kernel driver is loaded this file should exist, I can do a pre check if needed but I cant find any case where this file will be missing and we trigger retimer nvm update.

File: libfwupd/fwupd-enums.c (hunk 2 above)
Comment (krankun) [reply libfwupd/fwupd-enums.c]: done in https://github.com/fwupd/fwupd/pull/4046/commits/8d0d551cc1a62fbd54d246ec7a5ddb513bac95bd

File: libfwupd/fwupd-enums.h (hunk 1 above)
Comment (krankun) [reply libfwupd/fwupd-enums.h]: done in https://github.com/fwupd/fwupd/pull/4046/commits/8d0d551cc1a62fbd54d246ec7a5ddb513bac95bd

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 20 above)
Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: fixed in https://github.com/fwupd/fwupd/commit/98336139be858dd20f4089bfea421f0fda9031b1

File: libfwupdplugin/fu-device.c [hunk 22]
```diff
@@ -230,6 +230,8 @@ fu_device_internal_flag_to_string(FuDeviceInternalFlags flag)
 		return "use-parent-for-battery";
 	if (flag == FU_DEVICE_INTERNAL_FLAG_USE_PROXY_FALLBACK)
 		return "use-proxy-fallback";
+	if (flag == FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE)
```
Comment (superm1) [inline libfwupdplugin/fu-device.c:233]: Since moved to this file needs "internal" in flag name

File: libfwupdplugin/fu-device.h [hunk 23]
```diff
@@ -400,6 +400,14 @@ typedef guint64 FuDeviceInternalFlags;
  */
 #define FU_DEVICE_INTERNAL_FLAG_USE_PROXY_FALLBACK (1ull << 18)
 
+/**
+ * FWUPD_DEVICE_FLAG_NO_AUTO_REMOVE :
```
Comment (superm1) [inline libfwupdplugin/fu-device.h:404]: Since moved here needs internal in flag name

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 24]
```diff
@@ -319,6 +319,133 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, err)) {
+		if (!fu_udev_device_write_sysfs(device, offline, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting usb4 port offline failed");
+			return FALSE;
+		}
+	} else {
+		return FALSE;
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:350]: This particular one needs special handling - you need to return TRUE here

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 25]
```diff
@@ -319,6 +319,133 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device, GError **err)
+{
+	const gchar *offline = "usb4_port1/offline";
+	const gchar *rescan = "usb4_port1/rescan";
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, err)) {
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:341]: This particular one I wouldn't bother having an error argument or you need to pass a local error since you're going to return true if the file doesn't exist.

File: libfwupd/fwupd-enums.h [hunk 26]
```diff
@@ -493,6 +493,7 @@ typedef enum {
  * Since: 1.7.1
  */
 #define FWUPD_DEVICE_FLAG_AFFECTS_FDE (1llu << 45)
+
```
Comment (superm1) [inline libfwupd/fwupd-enums.h:496]: Don't change whitespace

File: libfwupdplugin/fu-device.c (hunk 22 above)
Comment (krankun) [reply libfwupdplugin/fu-device.c]: my bad fixed it

File: plugins/thunderbolt/fu-thunderbolt-device.c (hunk 25 above)
Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c:341]: fu_thunderbolt_check_usb4_port_path has err argument already !

Comment (superm1) [reply plugins/thunderbolt/fu-thunderbolt-device.c:341]: You need to pass a local error so you don't set error on true condition

Comment (krankun) [reply plugins/thunderbolt/fu-thunderbolt-device.c]: got it ok as done in https://github.com/fwupd/fwupd/commit/d4cca8d8eb125b4e86abc8b7cff2e7192bb664d3 ?

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 27]
```diff
@@ -319,6 +319,136 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	if (fu_thunderbolt_check_usb4_port_path(device, offline, err_usb4_port_offline_path)) {
+		if (!fu_udev_device_write_sysfs(device, offline, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting usb4 port offline failed");
+			return FALSE;
+		}
+	} else {
+		return TRUE;
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:352]: just a nit, you may want `g_debug("%s", err_usb4_port_offline_path->message);` here.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 28]
```diff
@@ -319,6 +319,136 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	if (fu_thunderbolt_check_usb4_port_path(device, rescan, err_usb4_port_rescan_path)) {
+		if (!fu_udev_device_write_sysfs(device, rescan, "1", err)) {
+			g_set_error_literal(err,
+					    FWUPD_ERROR,
+					    FWUPD_ERROR_INTERNAL,
+					    "Rescan on port failed");
+			return FALSE;
+		}
+	} else {
+		return TRUE;
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:363]: just a nit, you may want `g_debug("%s", err_usb4_port_rescan_path->message);` here.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 29]
```diff
@@ -319,6 +319,136 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+					    FWUPD_ERROR_INTERNAL,
+					    "Setting port online failed");
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static FuUdevDevice *
+get_udev_parent_device(FuDevice *device, GError **error)
```
Comment (superm1) [inline plugins/thunderbolt/fu-thunderbolt-device.c:388]: nit: prefix this with `fu_thunderbolt`

**Review by superm1 — COMMENTED** [review]
just some nits left to me.  @hughsie can you have a look again?

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 30]
```diff
@@ -319,6 +319,138 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	g_autoptr(GFile) file = g_file_new_for_path(fn);
+	if (!g_file_query_exists(file, NULL)) {
+		g_set_error(err, G_IO_ERROR, G_IO_ERROR_NOT_FOUND, "failed to find %s", fn);
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_set_port_offline(FuUdevDevice *device, GError **err)
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:337]: s/err/error please

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 31]
```diff
@@ -319,6 +319,138 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+{
+	const gchar *offline = "usb4_port1/offline";
+	const gchar *rescan = "usb4_port1/rescan";
+	GError *err_usb4_port_offline_path = NULL;
+	GError *err_usb4_port_rescan_path = NULL;
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:342]: /s/err_usb4_port_offline_path/error_offline please

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 32]
```diff
@@ -319,6 +319,138 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_thunderbolt_device_set_port_online(FuUdevDevice *device, GError **err)
+{
+	FuUdevDevice *udev = FU_UDEV_DEVICE(device);
+	const gchar *offline = "usb4_port1/offline";
+	GError **err_usb4_port_offline_path = NULL;
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:375]: same; and I think this needs to be g_autoptr

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 33]
```diff
@@ -319,6 +319,138 @@ fu_thunderbolt_device_get_attr_uint16(FuThunderboltDevice *self, const gchar *na
...
+		return NULL;
+	}
+	return fu_udev_device_new_with_context(fu_device_get_context(FU_DEVICE(self)),
+					       g_steal_pointer(&udev_parent2));
+}
+gboolean
+fu_thunderbolt_probe_retimer(FuDevice *device, GError **error)
+{
+	g_autoptr(FuUdevDevice) parent = NULL;
+	FuThunderboltDevice *self = FU_THUNDERBOLT_DEVICE(device);
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:434]: can you flip these the other way around please? just by convention self is first.

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 34]
```diff
@@ -422,6 +554,11 @@ fu_thunderbolt_device_setup_controller(FuDevice *device, GError **error)
 		fu_device_add_internal_flag(device, FU_DEVICE_INTERNAL_FLAG_REPLUG_MATCH_GUID);
 	}
 
+	if (self->device_type == FU_THUNDERBOLT_DEVICE_TYPE_HOST_CONTROLLER) {
+		if (fu_thunderbolt_device_set_port_offline(FU_UDEV_DEVICE(device), error))
+			return TRUE;
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:559]: what happens if it fails? surely you want:

	if (!fu_thunderbolt_device_set_port_offline(FU_UDEV_DEVICE(device), error))
			return FALSE;

File: plugins/thunderbolt/fu-thunderbolt-device.c [hunk 35]
```diff
@@ -763,10 +902,12 @@ fu_thunderbolt_device_write_firmware(FuDevice *device,
...
 	/* whether to wait for a device replug or not */
-	if (!fu_device_has_flag(device, FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE)) {
-		fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_RESTART);
+	if (fu_device_has_flag(device, FWUPD_DEVICE_FLAG_USABLE_DURING_UPDATE)) {
 		fu_device_set_remove_delay(device, FU_PLUGIN_THUNDERBOLT_UPDATE_TIMEOUT);
-		fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+		fu_progress_set_status(progress, FWUPD_STATUS_DEVICE_RESTART);
+		if (self->device_type != FU_THUNDERBOLT_DEVICE_TYPE_RETIMER) {
+			fu_device_add_flag(device, FWUPD_DEVICE_FLAG_WAIT_FOR_REPLUG);
+		}
```
Comment (hughsie) [inline plugins/thunderbolt/fu-thunderbolt-device.c:910]: no real need for braces

File: src/fu-udev-backend.c [hunk 36]
```diff
@@ -41,7 +41,8 @@ fu_udev_backend_device_remove(FuUdevBackend *self, GUdevDevice *udev_device)
 	/* find the device we enumerated */
 	device_tmp =
 	    fu_backend_lookup_by_id(FU_BACKEND(self), g_udev_device_get_sysfs_path(udev_device));
-	if (device_tmp != NULL) {
+	if (device_tmp != NULL &&
+	    !fu_device_has_flag(device_tmp, FWUPD_DEVICE_INTERNAL_FLAG_NO_AUTO_REMOVE)) {
```
Comment (hughsie) [inline src/fu-udev-backend.c:45]: i think you want the same thing in fu-**usb**-backend.c too.

Comment (krankun) [reply src/fu-udev-backend.c:45]: Sure added for  fu-usb-backend.c too  , what are the the benefits to add it to usb backend ?
for later platforms we want to extend this to multiport where both ports may or may not share same nvm payload, is there a place to keep a check on the extracted nvm files not to be cleaned up until all the ports nvm is updated ? currently we dont have this issue but just checking to extend the functionality for future...

Comment (hughsie) [reply src/fu-udev-backend.c:45]: > what are the the benefits to add it to usb backend

To make the codebase consistent. _NO_AUTO_REMOVE is the kind of thing I could see being used for USB devices too.

> , is there a place to keep a check on the extracted nvm files not to be cleaned up until all the ports nvm is updated

Back a bit; how do you mean? Surely each "port" would have a unique FuDevice and associated GUID?

**Comment by hughsie** [comment]
Thanks @superm1, much appreciated for all the reviews, and most of all for @krankun for all the fixes.

**Comment by krankun** [comment]
> Thanks @superm1, much appreciated for all the reviews, and most of all for @krankun for all the fixes.

Thanks for the reviews and support @superm1 @hughsie @djcampello @RajmohanMani
while this change serves the needs of current platforms we are good to go , in future we would need to keep supporting the few things below 
- Implement a generic code for  below three cases
1) two ports with two retimers share same nvm pay load { just update for one of them which is current case }
2) two ports with two retimers with each have their own nvm pay load { update nvm for both of them }
3) one port can have two retimers { if they share same pay load just update for one of them otherwise for both}

we will keep enhancing the tool for all the above three cases , so there is still some work ahead :)
